src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -531,10 +531,11 @@ void MarkMarshalSpec (IMarshalInfoProvider spec, in DependencyInfo reason, IMemb
 
 		void MarkCustomAttributes (ICustomAttributeProvider provider, in DependencyInfo reason, IMemberDefinition sourceLocationMember)
 		{
-			if (!provider.HasCustomAttributes)
+			if (!_context.CustomAttributes.HasAttributes (provider))
```

Comment:
Wouldn't be better to have `if (provider.HasCustomAttributes) { ... } ` for the section which iterates `provider.CustomAttributes`. The logic after that will work as it's